### PR TITLE
Add rule validation in AnomalyDetector constructor

### PR DESCRIPTION
**Pull request by kaituo**
### Description
This commit introduces rule validation within the AnomalyDetector constructor. Any validation errors are now propagated and displayed on the frontend to ensure immediate feedback.

Testing:
* Verified that validation errors are properly propagated and shown on the frontend.
* Added UTs to cover the new validation logic.

### Check List
- [X] New functionality includes testing.
- [X] New functionality has been documented.
- [X] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [X] Commits are signed per the DCO using `--signoff`.
- [X] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/anomaly-detection/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/timeseries/ml/MemoryAwareConcurrentHashmap.java [hunk 1]
```diff
@@ -55,48 +52,4 @@ public ModelState<RCFModelType> put(String key, ModelState<RCFModelType> value)
...
         return previousAssociatedState;
     }
-
-    /**
-     * Gets all of a config's model sizes hosted on a node
-     *
-     * @param configId config Id
-     * @return a map of model id to its memory size
-     */
-    public Map<String, Long> getModelSize(String configId) {
```
Comment (amitgalitz) [inline src/main/java/org/opensearch/timeseries/ml/MemoryAwareConcurrentHashmap.java:65]: where these just never used anymore?

File: src/main/java/org/opensearch/ad/model/AnomalyDetector.java [hunk 2]
```diff
@@ -720,4 +723,117 @@ private static Boolean onlyParseBooleanValue(XContentParser parser) throws IOExc
...
+     */
+    private void validateRules(List<Feature> features, List<Rule> rules) {
+        // Null check for rules
+        if (rules == null || rules.isEmpty()) {
+            return; // No suppression rules to validate; consider as valid
+        }
+
+        // Null check for features
+        if (features == null) {
+            // Cannot proceed with validation if features are null but rules are not null
```
Comment (amitgalitz) [inline src/main/java/org/opensearch/ad/model/AnomalyDetector.java:745]: nit: could add `|| features.isEmpty()` same we do for rules above

File: src/main/java/org/opensearch/ad/model/AnomalyDetector.java [hunk 3]
```diff
@@ -720,4 +723,117 @@ private static Boolean onlyParseBooleanValue(XContentParser parser) throws IOExc
...
+            this.errorMessage = "Suppression Rule Error: Features are not defined while suppression rules are provided.";
```
Comment (amitgalitz) [inline src/main/java/org/opensearch/ad/model/AnomalyDetector.java:746]: nit: "Suppression Rule Error" can be put in a constant

File: src/main/java/org/opensearch/ad/model/AnomalyDetector.java [hunk 4]
```diff
@@ -720,4 +723,117 @@ private static Boolean onlyParseBooleanValue(XContentParser parser) throws IOExc
...
+
+                // other threshold types may not have value operand
+                ThresholdType thresholdType = condition.getThresholdType();
+                if (thresholdType == ThresholdType.ACTUAL_OVER_EXPECTED_MARGIN
+                    || thresholdType == ThresholdType.EXPECTED_OVER_ACTUAL_MARGIN
+                    || thresholdType == ThresholdType.ACTUAL_OVER_EXPECTED_RATIO
+                    || thresholdType == ThresholdType.EXPECTED_OVER_ACTUAL_RATIO) {
+                    // Check if the value is not NaN
+                    double value = condition.getValue();
+                    if (Double.isNaN(value)) {
```
Comment (amitgalitz) [inline src/main/java/org/opensearch/ad/model/AnomalyDetector.java:815]: btw I will be changing value section to be okay with Null if operator type is ACTUAL_IS_OVER_EXPECTED or EXPECTED_IS_OVER_ACTUAL. Will make sure those aren't passed to RCF and seen as not a rule in that case later on too

**Review by amitgalitz — DISMISSED** [review]

File: src/main/java/org/opensearch/timeseries/ml/MemoryAwareConcurrentHashmap.java (hunk 1 above)
Comment (kaituo) [reply src/main/java/org/opensearch/timeseries/ml/MemoryAwareConcurrentHashmap.java:65]: yeah, they are not used after recent refactoring.

File: src/main/java/org/opensearch/ad/model/AnomalyDetector.java (hunk 2 above)
Comment (kaituo) [reply src/main/java/org/opensearch/ad/model/AnomalyDetector.java:745]: yeah, fixed.

File: src/main/java/org/opensearch/ad/model/AnomalyDetector.java (hunk 3 above)
Comment (kaituo) [reply src/main/java/org/opensearch/ad/model/AnomalyDetector.java:746]: fixed.

File: src/main/java/org/opensearch/ad/model/AnomalyDetector.java (hunk 4 above)
Comment (kaituo) [reply src/main/java/org/opensearch/ad/model/AnomalyDetector.java:815]: thanks.

**Review by amitgalitz — APPROVED** [review]

**Pull request merged**